### PR TITLE
Add a four-element AsyncLocalValueMap type

### PR DESCRIPTION
**Pull request by stephentoub**
We previously special-cased up to three active AsyncLocals in a given async flow, but it seems four is also very common.  Special-casing four as well results in four using ~20% less allocation and ~10% less CPU overhead in a simple mutation test.

```C#
private AsyncLocal<int> asyncLocal1 = new AsyncLocal<int>();
private AsyncLocal<int> asyncLocal2 = new AsyncLocal<int>();
private AsyncLocal<int> asyncLocal3 = new AsyncLocal<int>();
private AsyncLocal<int> asyncLocal4 = new AsyncLocal<int>();

[Benchmark(OperationsPerInvoke = 4000)]
public void Update()
{
    for (int i = 0; i < 1000; i++)
    {
        asyncLocal1.Value++;
        asyncLocal2.Value++;
        asyncLocal3.Value++;
        asyncLocal4.Value++;
    }
}
```

| Method |         Toolchain |     Mean | Ratio | Allocated | Alloc Ratio |
|------- |------------------ |---------:|------:|----------:|------------:|
| Update | \main\corerun.exe | 68.71 ns |  1.00 |     176 B |        1.00 |
| Update |   \pr\corerun.exe | 60.96 ns |  0.87 |     144 B |        0.82 |

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Threading/AsyncLocal.cs [hunk 1]
```diff
@@ -338,6 +407,12 @@ internal MultiElementAsyncLocalValueMap(int count)
                 _keyValues = new KeyValuePair<IAsyncLocal, object?>[count];
             }
 
+            internal MultiElementAsyncLocalValueMap(KeyValuePair<IAsyncLocal, object?>[] keyValues)
+            {
+                Debug.Assert(keyValues.Length is >= 5 and <= MaxMultiElements);
```
Comment (kouvel) [inline src/libraries/System.Private.CoreLib/src/System/Threading/AsyncLocal.cs:412]: Would it be beneficial for `MultiElementAsyncLocalValueMap.Set` to downgrade to a `FourElementAsyncLocalValueMap` when going from 5 to 4 elements?

**Review by kouvel — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Threading/AsyncLocal.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Threading/AsyncLocal.cs:412]: That was the intent, but I neglected to update this check:
https://github.com/dotnet/runtime/pull/68790/files#diff-f7c12a67956e72c9620ce862bff0cfe651e598ec1fa59abea4f7362c54603b1cR439
I'll fix that.

**Review by kouvel — APPROVED** [review]

**Comment by stephentoub** [comment]
Thanks for reviewing, @kouvel.

**Comment by davidfowl** [comment]
Activity, HttpContextAccessor, Logging scope, what's the 4th 😄. I need to go look...

**Pull request merged**